perf: Barriers, avoid network reader per component

paulpach opened the pull request:
@vis2k   you have seen this before.   This is the barrier patch updated to master.

I did some profiling to demonstrate why we may want this.   I spawned 1000 objects with 3 components.   They update their data every 1 second

Before:
<img width="604" alt="screen shot 2019-01-15 at 6 51 53 pm" src="https://user-images.githubusercontent.com/466007/51219675-7aef4280-18f7-11e9-937e-38f90c6b9ed1.png">
You can see 1.6M allocation and 10ms overhead 

After:
<img width="610" alt="screen shot 2019-01-15 at 6 53 30 pm" src="https://user-images.githubusercontent.com/466007/51219686-8fcbd600-18f7-11e9-9aa8-c4d29084127b.png">

The cost is down to 0.6M allocations and 5.39 ms.  

We could reduce it even further if we cache the network writer  (gasp,  the horror)

imerr commented:
Can confirm this is another huge performance improvement
Down to 0.6ms from 1.3ms now
Before: 
![image](https://user-images.githubusercontent.com/1426904/51222107-86189180-193c-11e9-9587-3b9a19cfa037.png)
After:
![image](https://user-images.githubusercontent.com/1426904/51222088-71d49480-193c-11e9-8e9b-82bdb6e03153.png)

vis2k commented:
can you guys share your benchmark? I would like to test one more approach too

paulpach commented:
@vis2k  Here is my benchmark
https://drive.google.com/open?id=1FV9RfqFWTSowKxFMOTqVZnbse7Gj_aNd

paulpach commented:
@imerr    Master has vis improvements.   This PR is up to date and so is 2018 branch.   Do you mind benchmarking all 3 again please?

imerr commented:
Master 
Server:  
![image](https://user-images.githubusercontent.com/1426904/52128898-74bacf00-2636-11e9-879e-133d66c01117.png) 5.3MB 50.9ms 44.05ms
Client: https://share.dl.je/2019/02/2019-02-01_15-32-36_9y5oYB1d8R.png
![image](https://user-images.githubusercontent.com/1426904/52129032-d0855800-2636-11e9-98cd-db1ed6268a93.png) 3.6MB 18.73ms 13.68ms
2018
Server: 
![image](https://user-images.githubusercontent.com/1426904/52129136-0c202200-2637-11e9-8295-a76ec62916d4.png) 0.7MB 18.16ms 16.72ms
Client: 
![image](https://user-images.githubusercontent.com/1426904/52129173-2eb23b00-2637-11e9-80f7-92e4b4eadb67.png) 1.9MB 14.31ms 11.7ms
#299 
Server: 
![image](https://user-images.githubusercontent.com/1426904/52129337-8ea8e180-2637-11e9-823c-89a7dad74e21.png) 5.3MB 47.99ms 41.44ms
Client: 
![image](https://user-images.githubusercontent.com/1426904/52129532-07a83900-2638-11e9-9b47-2d7d16aff88f.png) 2.2MB 13.60ms 10.52ms

imerr commented:
Here's also a table, had to leave earlier
Server:

| Branch | Time | Time incl. GC | Garbage |
|--------|-------|---------------|---------|
| Master | 44.05 | 50.9 | 5.3MB |
|  #299 | 41.44 | 47.99 |  5.3 |
| 2018 | 16.72 | 18.16 | 0.7MB |

Client: 

| Branch | Time | Time incl. GC | Garbage |
|--------|-------|---------------|---------|
| Master | 13.68 | 18.73 | 3.6MB |
|  #299 | 10.52 | 13.6 | 2.2MB |
| 2018 | 11.7 | 14.31 |  1.9MB |

vis2k commented:
will get back to this in a while. want to see real world impact with next 500+ CCU test

paulpach commented:
Ok.

This branch has mostly an impact on the client though.   Not sure what you expect to see in the 500+ CCU test.

Now 2018 branch should have a noticeable impact on the server

paulpach commented:
Ran some more profiling:

I added a monster prefab with 9 components to simulate modular ummorpg.  Each one updating something every second.
<img width="697" alt="Screen Shot 2019-03-23 at 10 05 58 PM" src="https://user-images.githubusercontent.com/466007/54874422-f4eced80-4db8-11e9-9e80-5973cec5edac.png">

I added 100 monsters to the scene.

I profiled synchronizing the monsters in master and with the barrier patch.

In master:
<img width="708" alt="Screen Shot 2019-03-23 at 10 05 34 PM" src="https://user-images.githubusercontent.com/466007/54874431-1221bc00-4db9-11e9-994c-bded75495762.png">

with barrier:
<img width="699" alt="Screen Shot 2019-03-23 at 10 10 06 PM" src="https://user-images.githubusercontent.com/466007/54874432-1c43ba80-4db9-11e9-93b8-fa5279e7a387.png">

paulpach commented:
To summarize the results:

| Branch | Time | Garbage |
|--------|-------|---------|
| Master | 3.05 | 333.5KB |
|  #299 | 2.12 |  123.7KB |

paulpach commented on Assets/Mirror/Runtime/NetworkIdentity.cs:
```diff
@@ -505,21 +505,17 @@ internal bool OnCheckObserver(NetworkConnection conn)
-        // -> OnSerialize writes length,componentData,length,componentData,...
-        // -> OnDeserialize carefully extracts each data, then deserializes each component with separate readers
-        //    -> it will be impossible to read too many or too few bytes in OnDeserialize
+        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```
```suggestion
```

rodolphito commented:
Is it just me or would the title of this PR make more sense as "Avoid Network *Reader* per component" ?

paulpach commented:
> Is it just me or would the title of this PR make more sense as "Avoid Network _Reader_ per component" ?

good point.

When I opened this pull request originally,  the goal was to eliminate the network writers per component.   That is already solved in master,  so the only thing left in this PR is to eliminate the network readers per component.

paulpach commented:
The allocation storm was addressed with 27b7e250a0451ae6a04222a9d035a5b0efdbeb99
